### PR TITLE
Cleaned up loading SfmData

### PR DESCRIPTION
**Pull request by dellaert**
- made static constructors
- deprecated functions using a non-const reference
- added an example in Python with Balbianello:
image

### CODE REVIEW

File: gtsam/sfm/SfmData.h [hunk 1]
```diff
@@ -61,6 +82,31 @@ struct SfmData {
...
+   *
+   * @param model a noise model for projection errors
+   * @param fixedCamera which camera to fix, if any (use boost::none if none)
+   * @param fixedPoint which point to fix, if any (use boost::none if none)
+   * @return NonlinearFactorGraph
+   */
+  NonlinearFactorGraph sfmFactorGraph(
+      const SharedNoiseModel& model = noiseModel::Isotropic::Sigma(2, 1.0),
+      boost::optional<size_t> fixedCamera = 0,
+      boost::optional<size_t> fixedPoint = 0) const;
```
Comment (akshay-krishnan) [inline gtsam/sfm/SfmData.h:108]: Nit, feel free to ignore if merging is urgent: 
I think it is unclear what the difference between generalSfmFactors and sfmFactorGraph is to a user from the name itself. Maybe sfmFactorGraphWithPriors is a better name?

**Review by akshay-krishnan — APPROVED** [review]
Looks good! I like the new static read methods, avoids quite some boilerplate.

**Comment by dellaert** [comment]
Thanks! The exception does not need to be flagged, it's just a deprecated method and making sure it returns a bool - that's a pretty useless behavior, that's also why deprecated.

File: gtsam/sfm/SfmData.cpp [hunk 2]
```diff
@@ -178,34 +179,34 @@ bool readBundler(const std::string &filename, SfmData &data) {
       track.siftIndices.emplace_back(cam_idx, point_idx);
     }
 
-    data.tracks.push_back(track);
+    sfmData.tracks.push_back(track);
   }
 
-  is.close();
```
[the remark this answers is not in the document]
Comment (dellaert) [reply gtsam/sfm/SfmData.cpp:184]: Ah, no, a stream closes when it is destroyed :-)

File: gtsam/sfm/SfmData.h (hunk 1 above)
Comment (dellaert) [reply gtsam/sfm/SfmData.h:108]: generalSfmFactors is sorta internal, so maybe I will move that to a different {} group

File: python/gtsam/utils/visual_data_generator.py [hunk 3]
```diff
@@ -36,7 +36,7 @@ def __init__(self, K=Cal3_S2(), nrCameras: int = 3, nrPoints: int = 4) -> None:
         self.cameras = [Pose3()] * nrCameras
         self.points = [Point3(0, 0, 0)] * nrPoints
 
-    def print_(self, s="") -> None:
+    def print(self, s="") -> None:
```
Comment (johnwlambert) [inline python/gtsam/utils/visual_data_generator.py:39]: Looks great. nit: type hint missing here for `s`

File: python/gtsam/tests/test_SfmData.py [hunk 4]
```diff
@@ -56,24 +62,58 @@ def test_data(self):
...
         self.data.addTrack(track2)
+
         # Number of tracks in SfmData is 2
         self.assertEqual(self.data.numberTracks(), 2)
+        
         # camera idx of first measurement of second track corresponds to i1
         cam_idx, img_measurement = self.data.track(1).measurement(0)
         self.assertEqual(cam_idx, i1)
 
+    def test_Balbianello(self):
```
Comment (johnwlambert) [inline python/gtsam/tests/test_SfmData.py:88]: nit: would be nice to add a 1-liner sentence explaining what the test is striving for

File: python/gtsam/tests/test_SfmData.py [hunk 5]
```diff
@@ -17,6 +17,7 @@
 import numpy as np
 
 import gtsam
+from gtsam import SfmData, SfmTrack, Point2, Point3
```
Comment (johnwlambert) [inline python/gtsam/tests/test_SfmData.py:20]: nit: sorted alphabetically, these should be 
```python
from gtsam import Point2, Point3, SfmData, SfmTrack
```

**Review by johnwlambert — APPROVED** [review]
LGTM, though a few minor python syntax changes could be made

**Comment by dellaert** [comment]
Okidoke!

**Comment by dellaert** [comment]
PS I know why the CI fails, will fix that as well.

File: python/gtsam/tests/test_SfmData.py (hunk 5 above)
Comment (dellaert) [reply python/gtsam/tests/test_SfmData.py:20]: Done

File: python/gtsam/utils/visual_data_generator.py (hunk 3 above)
Comment (dellaert) [reply python/gtsam/utils/visual_data_generator.py:39]: Done

**Pull request merged**